### PR TITLE
Baron's branch

### PR DESCRIPTION
**Pull request by JueWang268**
### Changes
- added keyboard shortcuts for all File menu items.
- handleExitMenuItem now triggered by clicking on the red dot on the top corner
- certain menu items are disabled when appropriate
- fixed textArea (to be replaced with code area) ratio; now it will always stretch to fit whole window

### Changes based on Prof's feedback
- no duplicate tabs allowed; user will be redirected to existing tab for the file they try to open again
- fixed handleCloseMenuItem and handleExitMenuItem. I restructured the if statement so it won't ask the user to save if no changes have been made to a non-empty file.

## Todo
- replace textAreas with VirtualizedScrollPanes and CodeAreas
- color brackets and keywords
- add keyword var
- exception handling on user's level
- improve the dirty detection

### CODE REVIEW

File: src/proj04BittingEllmerWang/Controller.java [hunk 1]
```diff
@@ -70,21 +89,34 @@ private void handleAboutMenuItem(Event event) {
...
         newTab.setId("Untitled-" + untitledNumber++);
-        newTab.setContent(new TextArea());
+        newTab.setContent(new VirtualizedScrollPane(new CodeArea()));
 
         // add new tab and move selection to front
         tabPane.getTabs().add(newTab);
         tabPane.getSelectionModel().select(newTab);
+
+        // re-enable the buttons when there are tabs
+        if (!tabPane.getTabs().isEmpty()) {
```
Comment (KitchenTable99) [inline src/proj04BittingEllmerWang/Controller.java]: This if statement isn't needed

File: src/proj04BittingEllmerWang/Controller.java [hunk 2]
```diff
@@ -70,21 +89,34 @@ private void handleAboutMenuItem(Event event) {
...
+            Close.setDisable(false);
```
Comment (KitchenTable99) [inline src/proj04BittingEllmerWang/Controller.java]: These methods could be combined into something like an enableAll method

File: src/proj04BittingEllmerWang/Controller.java [hunk 3]
```diff
@@ -110,24 +142,45 @@ private void handleOpenMenuItem(Event event) throws IOException {
...
 
+
         // if a valid file is selected
         if (selectedFile != null) {
             // get the path of the file selected
             String filePath = selectedFile.getPath();
+
+            // check if the file has already been opened in tabPane.
+            for (Object tabObj : tabPane.getTabs().toArray()) {
+                Tab tab = (Tab) tabObj;
```
Comment (KitchenTable99) [inline src/proj04BittingEllmerWang/Controller.java]: Can this be done by editing the for-each loop?

File: src/proj04BittingEllmerWang/Controller.java [hunk 4]
```diff
@@ -110,24 +142,45 @@ private void handleOpenMenuItem(Event event) throws IOException {
...
+            // set the content of the codeBox
+            codeBox.appendText(fileContent);
             // set the title of the tab
             String[] fileAncestors = filePath.split("/");
             currentTab.setText(fileAncestors[fileAncestors.length - 1]);
             currentTab.setId(filePath);
         }
+
+        // re-enable the buttons when there are tabs
+        if (!tabPane.getTabs().isEmpty()) {
```
Comment (KitchenTable99) [inline src/proj04BittingEllmerWang/Controller.java]: This code block is re-used, so there should definitely be an enableAll method. That method should be called if the method actually opens a new tab.

File: src/proj04BittingEllmerWang/Controller.java [hunk 5]
```diff
@@ -183,15 +236,27 @@ private void handleCloseMenuItem(Event event) {
...
+                }
             });
         }
         // if no changes have been made, the tab also closes
         else {
             tabPane.getTabs().remove(currentTab);
         }
+
+        // checks if there's any tab to close; if not, disable buttons
+        if (tabPane.getTabs().isEmpty()){
```
Comment (KitchenTable99) [inline src/proj04BittingEllmerWang/Controller.java]: disableAll method?

File: src/proj04BittingEllmerWang/Controller.java [hunk 6]
```diff
@@ -304,8 +377,8 @@ private boolean SaveFile(String content, File file) {
      *              and its source.
      */
     @FXML
-    private void handleExitMenuItem(Event event) {
-        while (tabPane.getSelectionModel().getSelectedItem() != null) {
+    protected void handleExitMenuItem(Event event) {
```
Comment (KitchenTable99) [inline src/proj04BittingEllmerWang/Controller.java]: why is this protected?

**Review by KitchenTable99 — CHANGES_REQUESTED** [review]
Added some comments about style and possible needed changes.

File: src/proj04BittingEllmerWang/Controller.java (hunk 6 above)
Comment (JueWang268) [reply src/proj04BittingEllmerWang/Controller.java]: Main, within the same package, needs to access it. I'm changing it to default.

**Review by KitchenTable99 — APPROVED** [review]
Looks good to me

**Pull request merged**